python_modules/libraries/dagster-dbt/dagster_dbt/core/resources_v2.py

```diff
@@ -978,66 +958,30 @@ def fetch_row_counts(
-            AssetCheckResult,
-        ]
-    ]:
-        logger.info("Row counts will be fetched for non-view models once they are materialized.")
-
-        # We keep a list of emitted Dagster events and pending futures which augment
-        # emitted events with additional metadata. This ensures we can yield events in the order
-        # they are emitted by dbt.
-        output_events_and_futures: List[Union[Future, DbtDagsterEventType]] = []
-
-        # Point at project directory to ensure dbt adapters run correctly
-        with pushd(str(self._dbt_cli_invocation.project_dir)), ThreadPoolExecutor(
-            max_workers=STREAM_EVENTS_THREADPOOL_SIZE
-        ) as executor:
-            self._dbt_run_thread = executor.submit(
-                self._stream_dbt_events_and_enqueue_postprocessing,
-                output_events_and_futures,
-                executor,
-            )
+            if isinstance(self._dbt_cli_invocation.adapter, DuckDBAdapter):
+                block_on_dbt_run = True
+        except ImportError:
+            pass
 
-            # Step through the list of output events and futures, yielding them in order
-            # once they are ready to be emitted
-            event_to_emit_idx = 0
-            while True:
-                all_work_complete = get_future_completion_state_or_err(
-                    [self._dbt_run_thread, *output_events_and_futures]
-                )
-                if all_work_complete and event_to_emit_idx >= len(output_events_and_futures):
-                    break
-
-                if event_to_emit_idx < len(output_events_and_futures):
-                    event_to_emit = output_events_and_futures[event_to_emit_idx]
-
-                    try:
-                        # If the next event to emit is a Future (waiting on postprocessing),
-                        # we need to wait for it to complete before yielding the event.
-                        event = (
-                            event_to_emit.result(timeout=0.1)
-                            if isinstance(event_to_emit, Future)
-                            else event_to_emit
-                        )
-                        yield event
-                        event_to_emit_idx += 1
-                    except:
-                        # If the Future has not completed, it will raise a TimeoutError.
-                        # Any other exception will be reraised in the main thread as part
-                        # of get_future_completion_state_or_err.
-                        pass
+        with pushd(str(self._dbt_cli_invocation.project_dir)):
+            return DbtEventIterator(
+                imap(
+                    executor=ThreadPoolExecutor(max_workers=num_threads),
+                    iterable=self,
+                    func=self._fetch_and_attach_row_count_metadata,
+                    block_on_enqueuing_task_completion=block_on_dbt_run,
+                    should_shutdown_executor=True,
+                ),
```

Comment:
Manually managing the context lifecycle of `ThreadPoolExecutor` seems suspect. Can we just add to this `with` block here instead?

Could we also name give the pool a thread name prefix to help us in debugging the future, if something happens?